### DISCUSSION TITLE
http://lotabout.github.io/2015/write-a-C-interpreter-2/

### DISCUSSION THREAD

**Discussion by lotabout** [lotabout/lotabout.github.io#9] [Announcements]
# http://lotabout.github.io/2015/write-a-C-interpreter-2/
手把手教你构建 C 语言编译器（2）- 虚拟机
http://lotabout.github.io/2015/write-a-C-interpreter-2/

**Comment by lotabout** [comment]
由 Patrick Wu 发布于 2018-12-22T11:27:11Z：

运行后 segmentation fault

**Reply by lotabout** [reply]
由 lotabout 发布于 2019-01-17T14:59:43Z：

因为代码里会有一些指针的强制转换，默认是 32 位的，可以尝试 \`gcc -m32\` 来编译

**Reply by lotabout** [reply]
由 Patrick Wu 发布于 2019-01-17T16:58:39Z：

是的 我后来发现了， 这个操作非常hacky :)

**Reply by lotabout** [reply]
由 unionx 发布于 2019-09-21T17:46:17Z：

用 -m32 编译之后好像也运行不了哦，你解决了吗？

**Reply by lotabout** [reply]
由 white jack 发布于 2020-04-21T09:30:31Z：

我也 segmentation fault  
後來發現是忘記 define int long long

**Comment by lotabout** [comment]
由 unionx 发布于 2019-09-21T16:53:37Z：

'jump if ax is zero' 这里重复了？应该是 is not zero？

**Comment by lotabout** [comment]
由 陳志遠 发布于 2020-04-24T06:57:43Z：

else if (op == SC) {ax = \*(char \*)\*sp++ = ax;} // save character to address, value in ax, address on stack   
↑ 想問版主這段代碼是否重複。

else if (op == SC) {\*(char \*)\*sp++ = ax;}

**Reply by lotabout** [reply]
由 lotabout 发布于 2020-04-29T01:28:10Z：

`*(char *)*sp++ = ax;` 的值最终要再赋值给 `ax`

**Reply by lotabout** [reply]
由 陳志遠 发布于 2020-05-08T15:03:46Z：

根據 = 賦值指令為右結合  
ex: a = b = a (最終都為a的值)  
else if (op == SC) {ax = \*(char \*)\*sp++ = ax;}  
ax的值依循右結合仍為ax，所以想問這樣寫是便於理解涵義還是我有語法上的誤解。  
感謝版主的解惑了。

**Comment by lotabout** [comment]
由 lanka Liu 发布于 2021-06-20T15:29:59Z：

運行後出現could not open((null))

**Reply by lotabout** [reply]
由 Zhixin Wang 发布于 2022-09-01T18:01:12Z：

因为这只是演示代码，并没有输入，所以把读取字符串那几行代码注释掉就可以了，不过注意要保留给src和old_src申请内存的那几行

**Comment by lotabout** [comment]
由 Robert Shannon 发布于 2023-09-23T06:33:51Z：

测试阶段好像少了给op赋值的操作吗？op = \*pc++; ?

**Comment by LeeeeeeM** [comment]
可惜了，只能看看思路，代码在macos 上执行不起来。
